brage-import/src/main/java/no/sikt/nva/brage/migration/lambda/BrageEntryEventConsumer.java

```diff
@@ -65,11 +67,74 @@ public BrageEntryEventConsumer() {
     public Publication handleRequest(S3Event s3Event, Context context) {
         return attempt(() -> parseBrageRecord(s3Event))
                    .map(publication -> pushAssociatedFilesToPersistedStorage(publication, s3Event))
+                   .map(publication -> hasExistingCristinIdentifier(publication)
+                                           ? attemptToUpdateExistingPublication(publication, s3Event)
+                                           : createNewPublication(publication, s3Event))
+                   .orElseThrow(fail -> handleSavingError(fail, s3Event));
+    }
+
+    private static Publication injectAssociatedArtifacts(Publication publication, Publication existingPublication) {
+        existingPublication.setAssociatedArtifacts(publication.getAssociatedArtifacts());
+        return existingPublication;
+    }
+
+    private static Publication getOnlyElement(List<Publication> publications) {
+        return Iterables.getOnlyElement(publications);
+    }
+
+    private boolean hasExistingCristinIdentifier(Publication publication) {
+        String cristinIdentifier = getCristinIdentifier(publication);
+        return nonNull(cristinIdentifier)
+               && !getPublicationsByCristinIdentifier(cristinIdentifier).isEmpty();
```

Comment:
You are querying the database with the same query atleast two times. Maybe cache the result instead?

Reply:
fikset